nixos/modules/services/torrent/deluge.nix

```diff
@@ -20,9 +60,93 @@ in {
             Number of files to allow deluged to open.
           '';
         };
+
+        coreConf = mkOption {
+          type = types.attrs;
+          example = {
+            download_location = "/srv/torrents/";
+            max_upload_speed = "1000.0";
+            share_ratio_limit = "2.0";
+            allow_remote = true;
+          };
+          description = ''
+            Deluge core configuration for the core.conf file. Only has
+            an effect when <option>services.deluge.declarative</option>
+            is set to <literal>true</literal>.
+            String values must be quoted, integer and boolean values must not.
+          '';
+        };
+
+      declarative = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Whether to use a declarative deluge configuration.
+          Only if set to <literal>true</literal>, the options
+          <option>services.deluge.coreConf</option>,
+          <option>services.deluge.listenPortRange</option> and
+          <option>services.deluge.auth</option> will be
+          applied.
+        '';
       };
 
-      deluge.web.enable = mkEnableOption "Deluge Web daemon";
+        port = mkOption {
+        type = types.ints.u16;
+          default = 58846;
+          description = ''
+            Deluge daemon port.
+          '';
+        };
+
+        listenPortRange = mkOption {
+          type = types.listOf types.attrs;
+          default = [ { from = 56881; to = 56889; } ];
+          description = ''
+            Set UDP listening port range used by DHT(IPv4, IPv6) and UDP tracker.
+            Only has an effect when <option>services.deluge.declarative</option>
+            is set to <literal>true</literal>.
+          '';
+        };
+
+        openFirewall = mkOption {
+          default = false;
+          type = types.bool;
+          description = "Whether to open the firewall for the specified ports.";
+        };
+
+        home = mkOption {
+          type = types.path;
+          default = "/var/lib/deluge";
+          description = ''
+            The directory where deluge will create files.
+          '';
+        };
+
+        authFile = mkOption {
+          type = types.nullOr types.path;
+          default = null;
```

Comment:
I recommend using `type = types.path` directly and not setting a default. A declarative setup requires setting this, and only in that case this option is actually used, so it can stay undefined in the stateful case. How it is now the `preStart` script will fail to evaluate when `declarative = true;` and `authFile` is unset. With this suggestion you get an error that "option was used but not defined".

Reply:
Yes I get this error in that case. Is it ok to leave that error as is, leaving the user finding why it fail?